Implement Power Outage Rollback Feature

SK8-infi opened the pull request:
## Description  
<!-- Provide a brief description of the changes made in this PR. -->  

Closes #917 

This PR implements a comprehensive Power Outage Rollback Feature for the krkn chaos engineering tool. The feature adds automatic rollback capabilities to restore powered-off nodes back to running state after a power outage scenario fails, making power outage scenarios more robust and reliable.

**Key Features:**
- Automatic rollback context management using `@set_rollback_context_decorator`
- Node restoration functionality via `rollback_shutdown_nodes()` method
- Support for all major cloud providers (AWS, GCP, Azure, OpenStack, IBM Cloud)
- Comprehensive error handling and partial failure recovery
- Detailed logging and status tracking for rollback operations
- Integration with existing krkn rollback framework

**Technical Implementation:**
- Enhanced `ShutDownScenarioPlugin` with rollback decorator and restore method
- Automatic rollback callable registration during cluster shutdown
- Proper cloud provider abstraction and timeout handling
- Follows krkn patterns and conventions for consistency

**Testing:**
- Added comprehensive test suite with 9 test cases
- Covers success scenarios, partial failures, error handling, and multi-cloud support
- Mock-based testing for reliable cloud provider interactions

This enhancement significantly improves the reliability of power outage chaos engineering operations by ensuring automatic recovery capabilities are always available.

## Documentation  
- [ ] **Is documentation needed for this update?**

## Related Documentation PR (if applicable)  
<!-- Add the link to the corresponding documentation PR in the website repository -->

SK8-infi commented:
Hi @paigerube14 
Happy new year 🎉 

Just checking in. Can you please review the PR again. Thanks

paigerube14 reviewed (commented):
@SK8-infi thanks so much again for this PR

I haven't had a chance to look back at this but am hitting this error now. Think with the location the rollback file get created it doesn't have krkn. Also saw the functional tests aren't passing if you could take a look, this PR would be much appreciated  


```
1111-ck5s7-worker-c-8mv7f']
2026-03-23 15:30:36,844 [ERROR] Failed to rollback shutdown nodes: No module named 'krkn'
Traceback (most recent call last):
  File "/tmp/kraken-rollback/1774294141730584000-70b51d0b-e8a9-4d14-a73c-75cae2bdfc3f/cluster_shut_down_scenarios_1774294138393881000_mksjbrrc.py", line 146, in <module>
    rollback_shutdown_nodes(
    ~~~~~~~~~~~~~~~~~~~~~~~^
        rollback_content,
        ^^^^^^^^^^^^^^^^^
        lib_telemetry
        ^^^^^^^^^^^^^
    )
    ^
  File "/tmp/kraken-rollback/1774294141730584000-70b51d0b-e8a9-4d14-a73c-75cae2bdfc3f/cluster_shut_down_scenarios_1774294138393881000_mksjbrrc.py", line 53, in rollback_shutdown_nodes
    from krkn.scenario_plugins.node_actions.gcp_node_scenarios import GCP
ModuleNotFoundError: No module named 'krkn'
```

SK8-infi commented:
Sure. Will look into it this week 👍

SK8-infi committed:
Implement Power Outage Rollback Feature

Signed-off-by: SK8-infi <shivansh.katiyar1712@gmail.com>

SK8-infi commented:
Hii @paigerube14 
Implemented skip_kubernetes + structured cloud_type/instance_ids on RollbackContent, wired shutdown rollback to use it, pass None telemetry for cloud-only rollback, skip kube init in the generated rollback script, and add sys.path/KRKN_ROOT so rollback files can import krkn from temp dirs; updated tests

paigerube14 commented:
hi @SK8-infi, I tried running this PR on my gcp cluster and I'm hitting the below error, can you please take a look at how we could fix this


```
% python  /Users/prubenda/.krkn/rollback/1778772679761771000-da366100-14c2-45c4-bddd-27c16f2fbae2/cluster_shut_down_scenarios_1778772675462890000_tubjbmki.py
2026-05-14 11:32:12,545 [INFO] Executing rollback callable...
2026-05-14 11:32:12,545 [INFO] Rolling back shutdown for 6 nodes on gcp
2026-05-14 11:32:12,545 [INFO] Node IDs: ['prubenda1111-ck5s7-master-0', 'prubenda1111-ck5s7-master-1', 'prubenda1111-ck5s7-master-2', 'prubenda1111-ck5s7-worker-a-8kp4k', 'prubenda1111-ck5s7-worker-b-tz26t', 'prubenda1111-ck5s7-worker-c-8mv7f']
/Users/prubenda/Github/pplrepo/SK8-infi/krkn/venv/lib/python3.11/site-packages/paramiko/pkey.py:100: CryptographyDeprecationWarning: TripleDES has been moved to cryptography.hazmat.decrepit.ciphers.algorithms.TripleDES and will be removed from cryptography.hazmat.primitives.ciphers.algorithms in 48.0.0.
  "cipher": algorithms.TripleDES,
/Users/prubenda/Github/pplrepo/SK8-infi/krkn/venv/lib/python3.11/site-packages/paramiko/transport.py:259: CryptographyDeprecationWarning: TripleDES has been moved to cryptography.hazmat.decrepit.ciphers.algorithms.TripleDES and will be removed from cryptography.hazmat.primitives.ciphers.algorithms in 48.0.0.
  "class": algorithms.TripleDES,
2026-05-14 11:32:13,819 [INFO] Starting instances for rollback...
2026-05-14 11:32:14,430 [ERROR] Failed to start node instance ['prubenda1111-ck5s7-master-0', 'prubenda1111-ck5s7-master-1', 'prubenda1111-ck5s7-master-2', 'prubenda1111-ck5s7-worker-a-8kp4k', 'prubenda1111-ck5s7-worker-b-tz26t', 'prubenda1111-ck5s7-worker-c-8mv7f']. Encountered following exception: bad argument type for built-in operation.
2026-05-14 11:32:14,430 [ERROR] Failed to start instances: 
2026-05-14 11:32:14,431 [ERROR] Failed to rollback shutdown nodes: 
Traceback (most recent call last):
  File "/Users/prubenda/Github/pplrepo/SK8-infi/krkn/krkn/scenario_plugins/node_actions/gcp_node_scenarios.py", line 82, in start_instances
    request = compute_v1.StartInstanceRequest(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/prubenda/Github/pplrepo/SK8-infi/krkn/venv/lib/python3.11/site-packages/proto/message.py", line 683, in __init__
    super().__setattr__("_pb", self._meta.pb(**params))
                               ^^^^^^^^^^^^^^^^^^^^^^^
TypeError: bad argument type for built-in operation

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/prubenda/.krkn/rollback/1778772679761771000-da366100-14c2-45c4-bddd-27c16f2fbae2/cluster_shut_down_scenarios_1778772675462890000_tubjbmki.py", line 167, in <module>
    rollback_shutdown_nodes(
  File "/Users/prubenda/.krkn/rollback/1778772679761771000-da366100-14c2-45c4-bddd-27c16f2fbae2/cluster_shut_down_scenarios_1778772675462890000_tubjbmki.py", line 91, in rollback_shutdown_nodes
    cloud_object.start_instances(node_ids)
  File "/Users/prubenda/Github/pplrepo/SK8-infi/krkn/krkn/scenario_plugins/node_actions/gcp_node_scenarios.py", line 95, in start_instances
    raise RuntimeError()
RuntimeError
```

paigerube14 commented:
Please take a look at the failing unit tests 

```
======================================================================
FAIL: test_run_success_aws (test_shut_down_scenario_plugin.TestShutDownScenarioPlugin.test_run_success_aws)
Test successful run of shut down scenario with AWS cloud type
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.11.15/x64/lib/python3.11/unittest/mock.py", line 1378, in patched
    return func(*newargs, **newkeywargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/krkn/krkn/tests/test_shut_down_scenario_plugin.py", line 67, in test_run_success_aws
    result = self.plugin.run(
             ^^^^^^^^^^^^^^^^
  File "/home/runner/work/krkn/krkn/krkn/rollback/handler.py", line 69, in wrapper
    assert run_uuid is not None, "run_uuid must be provided in kwargs"
           ^^^^^^^^^^^^^^^^^^^^
AssertionError: run_uuid must be provided in kwargs

======================================================================
FAIL: test_run_with_exception (test_shut_down_scenario_plugin.TestShutDownScenarioPlugin.test_run_with_exception)
Test run handles exceptions and returns 1
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.11.15/x64/lib/python3.11/unittest/mock.py", line 1378, in patched
    return func(*newargs, **newkeywargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/krkn/krkn/tests/test_shut_down_scenario_plugin.py", line 85, in test_run_with_exception
    result = self.plugin.run(
             ^^^^^^^^^^^^^^^^
  File "/home/runner/work/krkn/krkn/krkn/rollback/handler.py", line 69, in wrapper
    assert run_uuid is not None, "run_uuid must be provided in kwargs"
           ^^^^^^^^^^^^^^^^^^^^
AssertionError: run_uuid must be provided in kwargs

----------------------------------------------------------------------
```

paigerube14 commented:
The rollback itself seems to be working properly but getting this error at the end of starting the nodes
```
2026-06-01 09:49:55,141 [ERROR] Failed to rollback shutdown nodes: name 'time' is not defined
2026-06-01 09:49:55,141 [ERROR] Failed to execute rollback version file /Users/prubenda/.krkn/rollback/1780319961749741000-e02e6077-2860-4c0e-bc13-92f869163540/cluster_shut_down_scenarios_1780319956983522000_sodqkwsa.py: name 'time' is not defined
Traceback (most recent call last):
  File "/Users/prubenda/Github/pplrepo/SK8-infi/krkn/run_kraken.py", line 867, in <module>
    retval = main(options, command)
             ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/prubenda/Github/pplrepo/SK8-infi/krkn/run_kraken.py", line 450, in main
    scenario_plugin.run_scenarios(
  File "/Users/prubenda/Github/pplrepo/SK8-infi/krkn/krkn/scenario_plugins/abstract_scenario_plugin.py", line 146, in run_scenarios
    execute_rollback_version_files(
  File "/Users/prubenda/Github/pplrepo/SK8-infi/krkn/krkn/rollback/handler.py", line 178, in execute_rollback_version_files
    rollback_callable(rollback_content, telemetry_arg)
  File "/Users/prubenda/.krkn/rollback/1780319961749741000-e02e6077-2860-4c0e-bc13-92f869163540/cluster_shut_down_scenarios_1780319956983522000_sodqkwsa.py", line 141, in rollback_shutdown_nodes
    time.sleep(60)  # Shorter wait for rollback scenario
    ^^^^
NameError: name 'time' is not defined
```

SK8-infi commented:
Okay wait. Let me check.

SK8-infi commented:
Hii @paigerube14 
- added import time inside rollback_shutdown_nodes so newly generated rollback files include it for time.sleep(60)
Or should I change the version template to include the time library? But felt this was enough too for now

- rollback decorator now accepts run_uuid from positional args so shut_down unit tests pass

paigerube14 commented:
Power outage rollback working for me on a gcp cluster 

```
 % python /tmp/1780603106757356000-14ebbedc-ecd0-4541-a052-3048e3cbefe7/cluster_shut_down_scenarios_1780603101604453000_wzssdenl.py
2026-06-04 16:01:05,112 [INFO] Executing rollback callable...
2026-06-04 16:01:05,112 [INFO] Rolling back shutdown for 6 nodes on gcp
2026-06-04 16:01:05,112 [INFO] Node IDs: ['***-master-0', '***-master-1', '***-master-2', '***-worker-a-8kp4k', '***-worker-b-tz26t', '***-worker-c-8mv7f']
2026-06-04 16:01:06,869 [INFO] Starting instances for rollback...
2026-06-04 16:01:06,869 [INFO] Starting instance for rollback: ***-master-0
2026-06-04 16:01:07,964 [INFO] Instance: ***-master-0 started
2026-06-04 16:01:07,964 [INFO] Starting instance for rollback: ***-master-1
2026-06-04 16:01:08,795 [INFO] Instance: ***-master-1 started
2026-06-04 16:01:08,796 [INFO] Starting instance for rollback: ***-master-2
2026-06-04 16:01:09,679 [INFO] Instance: ***-master-2 started
2026-06-04 16:01:09,679 [INFO] Starting instance for rollback: ***-worker-a-8kp4k
2026-06-04 16:01:10,670 [INFO] Instance: ***-worker-a-8kp4k started
2026-06-04 16:01:10,670 [INFO] Starting instance for rollback: ***-worker-b-tz26t
2026-06-04 16:01:11,607 [INFO] Instance: ***-worker-b-tz26t started
2026-06-04 16:01:11,608 [INFO] Starting instance for rollback: ***-worker-c-8mv7f
2026-06-04 16:01:12,698 [INFO] Instance: ***-worker-c-8mv7f started
2026-06-04 16:01:12,698 [INFO] Waiting for instances to be running...
2026-06-04 16:01:12,698 [INFO] Waiting for node ***-master-0 to be running...
2026-06-04 16:01:13,305 [INFO] Status of instance ***-master-0: STAGING
2026-06-04 16:01:18,988 [INFO] Status of instance ***-master-0: RUNNING
2026-06-04 16:01:18,988 [INFO] status matches, endRUNNINGRUNNING
2026-06-04 16:01:18,988 [INFO] Successfully restored node: ***-master-0
2026-06-04 16:01:18,989 [INFO] Waiting for node ***-master-1 to be running...
2026-06-04 16:01:19,692 [INFO] Status of instance ***-master-1: RUNNING
2026-06-04 16:01:19,693 [INFO] status matches, endRUNNINGRUNNING
2026-06-04 16:01:19,693 [INFO] Successfully restored node: ***-master-1
2026-06-04 16:01:19,693 [INFO] Waiting for node ***-master-2 to be running...
2026-06-04 16:01:20,343 [INFO] Status of instance ***-master-2: STAGING
2026-06-04 16:01:25,936 [INFO] Status of instance ***-master-2: STAGING
2026-06-04 16:01:31,584 [INFO] Status of instance ***-master-2: RUNNING
2026-06-04 16:01:31,585 [INFO] status matches, endRUNNINGRUNNING
2026-06-04 16:01:31,585 [INFO] Successfully restored node: ***-master-2
2026-06-04 16:01:31,585 [INFO] Waiting for node ***-worker-a-8kp4k to be running...
2026-06-04 16:01:32,056 [INFO] Status of instance ***-worker-a-8kp4k: RUNNING
2026-06-04 16:01:32,056 [INFO] status matches, endRUNNINGRUNNING
2026-06-04 16:01:32,056 [INFO] Successfully restored node: ***-worker-a-8kp4k
2026-06-04 16:01:32,056 [INFO] Waiting for node ***-worker-b-tz26t to be running...
2026-06-04 16:01:32,596 [INFO] Status of instance ***-worker-b-tz26t: RUNNING
2026-06-04 16:01:32,597 [INFO] status matches, endRUNNINGRUNNING
2026-06-04 16:01:32,597 [INFO] Successfully restored node: ***-worker-b-tz26t
2026-06-04 16:01:32,597 [INFO] Waiting for node ***-worker-c-8mv7f to be running...
2026-06-04 16:01:34,317 [INFO] Status of instance ***-worker-c-8mv7f: RUNNING
2026-06-04 16:01:34,317 [INFO] status matches, endRUNNINGRUNNING
2026-06-04 16:01:34,317 [INFO] Successfully restored node: ***-worker-c-8mv7f
2026-06-04 16:01:34,317 [INFO] Rollback completed successfully for all 6 nodes
2026-06-04 16:01:34,317 [INFO] Waiting for cluster components to initialize after rollback...
2026-06-04 16:02:34,321 [INFO] Rollback of shutdown nodes completed.
2026-06-04 16:02:34,330 [INFO] Rollback completed.
```

paigerube14 commented:
@SK8-infi thanks so much for the work on this PR! are you able to squash your commits into 1 and we can get this merged! This is a huge add to krkn and helping us rollback a cluster that is fully down

SK8-infi commented:
Hii @paigerube14 . Squashed everything into one commit. 

Well this was one of the most challenging issues I took on. Learned a lot working on it.
Is there any other similar issue currently stale I may work on?? Would love to continue..
